### PR TITLE
dns: Add `AppleDnsResolverConfig.include_unroutable_families` config

### PR DESCRIPTION
**Pull request by jpsim**
Commit Message: dns: Add `AppleDnsResolverConfig.include_unroutable_families` config
Additional Description: If set, the resolver will avoid the system's heuristics to only return IPv4 or IPv6 addresses that it considers to be "routable", instead returning all possible IPv4 or IPv6 addresses. This setting is ignored if the DNS lookup family is set to v4-only or v6-only. This may be a useful setting to specify if the addresses considered unroutable by the system's heuristics may in practice be routable.
Risk Level: Low, off by default, only impacts Apple platforms.
Testing: Added unit tests. I've confirmed both branches of the code paths are hit when the flag is on and off via Envoy Mobile on a physical iPhone.
Docs Changes: Proto docs added
Release Notes: Added
Platform Specific Features: Apple-platforms only

### CODE REVIEW

**Comment by jpsim** [comment]
This mostly exposing a way to configure the behavior we had prior to https://github.com/envoyproxy/envoy/pull/18418.

**Comment by htuch** [comment]
>  if the addresses considered unroutable by the system's heuristics may in practice be routable.

Can you elaborate on this? Would be curious to understand if the issue is the heuristics or that we probably should fundamentally have the escape hatch because heuristics are heuristics. CC @mattklein123 (who owns the Apple DNS stuff since @junr03 is no longer maintainer?).

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks makes sense to me at a high level.

/wait

File: api/envoy/extensions/network/dns_resolver/apple/v3/apple_dns_resolver.proto
Comment (htuch) [reply api/envoy/extensions/network/dns_resolver/apple/v3/apple_dns_resolver.proto:22]: Yes, definitely, this totally makes sense with the mention of Happy Eyeballs.

**Comment by goaway** [comment]
Just to drop a clarifying comment on here - we generally believe that having this heuristic in place on iOS should be an improvement, at least based on its description, but nonetheless, we _have_ observed cases in practice where the Apple DNS API returns a single non-functional IP when we know that we've advertised both versions and the other version is the one that actually works.

Being able to quantify its impact would be useful, and also being able to run an experiment where we let Happy Eyeballs take precedence in selecting the preferred IP seems worth trying. And since Apple left the value configurable, it also seems reasonable to allow it to be configured through EM's interface as well.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Review by htuch — APPROVED** [review]
LGTM, will defer to @mattklein123 to merge.

**Comment by jpsim** [comment]
Thanks for the reviews and the discussion. I'll try to include more context up front in the future.

**Pull request merged**
